hw4/04-templates.js

```diff
@@ -1,62 +1,171 @@
-  });
-});
+const url = "https://restcountries.com/v3.1/all";
 
-app.get('/capitals', (req, res) => {
-  // map the output array to create an array with country names and capitals
-  // check for empty data in the output array
+//Format response from countries API and return an array containing the requested values.
+function parseData(data) {
+  let countries = [];
+  for (item of data) {
+    name = item.name.common; //Get each country's common name
+    pop = item.population.toLocaleString(); //Get each country's pop and format the value
+    capital = item.capital;
+    region = item.region;
+    let str = `${name};${pop};${capital};${region}`;
+    countries.push(str); //Add string to array
+  }
 
-  let countries = ['Afghanistan', 'Aland Islands', 'Albania'];
+  console.log(countries);
+  return countries;
+}
 
-  res.render('page', {
-    heading: 'Countries and Capitals',
-    results: countries,
+app.get("/", (req, res) => {
+  res.render("index", {
+    heading: "Countries of the World",
+    main: "Welcome to this application. Using the REST Countries API, we will be showing the countries and capitals of the world, the most populous countries in the world, and the number of countries in each region of the world",
   });
 });
 
-app.get('/populous', (req, res) => {
-  // filter the output array for the countries with population of 50 million or more
-  // sort the resulting array to show the results in order of population
-  // map the resulting array into a new array with the country name and formatted population
 
-  let populous = ['China', 'India', 'United States of America'];
 
-  res.render('page', {
-    heading: 'Most Populous Countries',
-    results: populous,
-  });
+app.get("/capitals", async (req, res) => {
+  try { //Get the data from the server
+    let response = await axios({
+      url: url,
+      method: "GET",
+      timeout: 8_000,
+      headers: {
+        "Content-Type": "text/html",
+      },
+    });
+    if (response.status == 200) { //Got data- format and display according to the requirements of the route.
+      let countries = [];
+      let str;
+      let data = parseData(response.data);
+      data.sort();
+      for (item of data) {
+        let elem = item.split(";");
+        let nocap = `no data`;
+        elem[2] === "undefined"
+          ? (str = `${elem[0]} - ${nocap}`)
+          : (str = `${elem[0]} - ${elem[2]}`);
+        countries.push(str);
+      }
+      res.render("page", {
+        heading: "Capitals of the World",
+        results: countries,
+      });
+    }
+    return response.data;
+  } catch (err) {
+    console.log(err);
+  }
 });
 
-app.get('/regions', (req, res) => {
-  // reduce the output array in a resulting object that will feature the numbers of countries in each region
-  // disregard empty data from the output array
+app.get("/populous", async (req, res) => {
+  try {//Get the data from the server
+    let response = await axios({
```

Comment:
don't call the  country api for each api ex capital or region, just load the whole data and serve for every api

Reply:
That was my initial idea- a function that was called when the window loaded, but I couldn't get the async to work.  I will try to play with it some more this weekend.